### ISSUE TITLE
restore fails with "Nothing to do. This project does not specify any packages for NuGet to restore."

### ISSUE THREAD

**Issue by joelmartinez** [NuGet/Home#6331]
## Details about Problem

I'm trying to get the absolute simplest example of a `.proj` file that restores some nuget packages going. Ideally, this would let me use `PackageReference` instead of having a separate `packages.config` file. But it doesn't seem to be working ... not sure if I'm missing something in the .proj file or what (_edit: seems like the verbose logs below might offer a clue_).

The ultimate goal for me is to use _msbuild_ as a cross-platform build automation tool. Part of my process involves downloading some dependencies from nuget.

- NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): **nuget.exe with a `.proj` file using `PackageReference`**
- NuGet version (x.x.x.xxx): **4.3.0.4406**
- Mono version: **5.4.1.7**
- OS version (i.e. win10 v1607 (14393.321)): **macOS Sierra 10.12.6** 

## Detailed repro steps so we can see the same problem

1. Create `build.proj` file with the following contents
```xml
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <ItemGroup>
        <PackageReference Include="mdoc" Version="5.3.5.2"  />
    </ItemGroup>
</Project>
```

2. run the command `nuget restore build.proj -PackagesDirectory Packages`. On my machine, it results in the following output:
```
MSBuild auto-detection: using msbuild version '15.0' from '/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/msbuild/15.0/bin/'.
Nothing to do. This project does not specify any packages for NuGet to restore.
```

### Verbose Logs

```bash
$ nuget restore build.proj -PackagesDirectory Packages -verbosity detailed
NuGet Version: 4.3.0.4406
MSBuild P2P timeout [ms]: 120000
MSBuild auto-detection: using msbuild version '15.0' from '/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/msbuild/15.0/bin/'. Use option -MSBuildVersion to force nuget to use a specific versionof MSBuild.
/Library/Frameworks/Mono.framework/Versions/Current/Commands/msbuild "/var/folders/wv/scnhx5js2slfm_hp76w1nq_40000gn/T/NuGetScratch/z0d88v94.0uo.nugetinputs.targets" /t:GenerateRestoreGraphFile /nologo /nr:false /v:q /p:NuGetRestoreTargets="/var/folders/wv/scnhx5js2slfm_hp76w1nq_40000gn/T/NuGetScratch/c59y6y5r.x9l.nugetrestore.targets" /p:RestoreTaskAssemblyFile="/Library/Frameworks/Mono.framework/Versions/5.4.1/lib/mono/nuget/nuget.exe" /p:RestorePackagesPath="Packages"
/Users/jmartinez/dev/xamarin/Xamarin.Forms-api-docs/build.proj : error MSB4057: The target "_IsProjectRestoreSupported" does not exist in the project.
/var/folders/wv/scnhx5js2slfm_hp76w1nq_40000gn/T/NuGetScratch/c59y6y5r.x9l.nugetrestore.targets(231,5):warning : Skipping restore for project '/Users/jmartinez/dev/xamarin/Xamarin.Forms-api-docs/build.proj'. The project file may be invalid or missing targets required for restore. [/var/folders/wv/scnhx5js2slfm_hp76w1nq_40000gn/T/NuGetScratch/z0d88v94.0uo.nugetinputs.targets]
/var/folders/wv/scnhx5js2slfm_hp76w1nq_40000gn/T/NuGetScratch/c59y6y5r.x9l.nugetrestore.targets(131,5):warning : Unable to find a project to restore! [/var/folders/wv/scnhx5js2slfm_hp76w1nq_40000gn/T/NuGetScratch/z0d88v94.0uo.nugetinputs.targets]

Nothing to do. This project does not specify any packages for NuGet to restore.
```

Hmm ... perhaps the verbose logs suggest a cause. Is there a way to keep the dependencies in the msbuild file, or will I have to fall back to using `packages.config` ?

**Comment by mishra14** [comment]
@joelmartinez You need to create a project that imports the targets needed to by nuget restore.

Have you tried [dotnet cli](https://www.microsoft.com/net/download/windows)? dotnet cli has a [command ](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-new?tabs=netcore2x)that lets you create cross-platform projects.

For instance - 

`dotnet new console`
`dotnet add package mdoc --version 5.3.5.2`
`dotnet build`

Please let us know if you are still blocked.

**Comment by joelmartinez** [comment]
@mishra14 thanks for the information. I _have_ tried `dotnet cli`, but it doesn't really match my intended use case. I'm not *necessarily* trying to compile some code (yes, that's often a sub-component of what I'm automating, but not the primary case). Your example above makes the following `.csproj`

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="mdoc" Version="5.3.5.2" />
  </ItemGroup>
</Project>
```

In my case, as I mentioned, I'm not (necessarily) trying to output an executable or library. I want to fetch some dependencies via nuget (CLI tools), and then invoke them in various ways ... for CI integration specifically.

What targets need to be imported for nuget restore to do the right thing with the `.proj` file?

**Comment by mishra14** [comment]
`Microsoft.Common.props` contain restore targets that you want. Will that work?

**Comment by joelmartinez** [comment]
@mishra14 I added `<Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" />`, but unfortunately I'm still getting the same message. Perhaps I've misunderstood your suggestion ... can you expand on your answer a bit?

fwiw, I really appreciate any additional info you can help with :)

**Comment by mishra14** [comment]
@joelmartinez I thought about your scenario a bit and realized that you just need a `packages.config` file - 
```
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="mdoc" version="5.3.5.2" />
</packages>
```

Then you need to run `nuget restore packages.config -PackagesDirectory Packages`

This will download the package into `.\Packages` folder.

Let me know if this is not the right solution for you. :)

**Comment by joelmartinez** [comment]
@mishra14 ok ... I was hoping to avoid having a separate file, as I'm trying to come up with a bare minimum example to add to my documentation; the `PackageReference` looked like it would fit my scenario perfectly. But it seems like there's only shallow support for it in some scenarios. As you mentioned, having a `packages.config` file does indeed work, so I'm going to move forward with that.

*follow up question* ... according to [the documentation](https://docs.microsoft.com/en-us/nuget/schema/msbuild-targets#restore-target), I should be able to call `MSBuild build.proj /t:restore` and have it restore the project. However, when I try that (with `packages.config` present), I get

```
$ msbuild build.proj /t:restore
Microsoft (R) Build Engine version 15.4.0.0 (master/5bb58816 Tue Oct  3 18:58:28 EDT 2017) for Mono
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 12/21/2017 1:43:41 PM.
Project "/Users/jmartinez/dev/github/mdoc-docfx/build.proj" on node 1 (restore target(s)).
/Users/jmartinez/dev/github/mdoc-docfx/build.proj : error MSB4057: The target "restore" does not exist in the project.
Done Building Project "/Users/jmartinez/dev/github/mdoc-docfx/build.proj" (restore target(s)) -- FAILED.

Build FAILED.
```

The only way I can get it to work is to include the following target in my project:
```xml
<Target Name="restore">
    <Exec Command="nuget restore build.proj -PackagesDirectory Packages" />
</Target>
```
it works, but seems awfully circular ... can you help me understand why `msbuild /t:restore` doesn't work in my scenario, without the additional target? remember, I'm trying to simplify this example to the bare bones required to make it work. Thanks!

**Comment by joelmartinez** [comment]
In case the additional context helps ... I'm looking to replace my use of a `Makefile` ([current here](https://github.com/joelmartinez/mdoc-docfx/blob/master/Makefile)), with using msbuild to orchestrate the running of these tools. This would make it an easier cross platform solution than relying on _make_.

**Comment by emgarten** [comment]
The restore target must exist. As mentioned earlier you will need to reference this common target to get restore to work.

**Comment by sim590** [comment]
Hi. I'm a bit late to the party, but it seems it is feasible to import `Microsoft.Common.targets` as follows:
```xml
<Import Project="$(MSBuildBinPath)\Microsoft.Common.targets" />
```
and that enables the following command to work on any `.proj` file that does contain the above line:
```
msbuild /target:restore my_proj_file.proj
```
I hope it helps.

### On previous answers

At first, `Microsoft.Common.props` was proposed to be imported, but it doesn't import the targets.

>The restore target must exist. As mentioned earlier you will need to reference this common target to get restore to work.

Here, it seems like @emgarten meant what I explained above, but he wrote it in plain English which I didn't think was that obvious, but I'm not a big Microsoft savvy guy, so...
